network/gossip.go

```diff
@@ -22,10 +23,13 @@ const (
 type Topic struct {
 	logger hclog.Logger
 
-	topic   *pubsub.Topic
-	typ     reflect.Type
-	closeCh chan struct{}
-	closed  *uint64
+	topic  *pubsub.Topic
+	typ    reflect.Type
+	closed *uint64
+
+	ctx       context.Context
+	cancelFn  context.CancelFunc
```

Comment:
It would be better to preserve `closeCh` to signal cancellation since having a `context.Context` captured within the state is considered a bad practice (didn't know myself either, until Evgeny pointed that out on one occasion in one of the mine PRs). 

 Generally, `context` is meant to be created when needed locally, and propagated in the call chain via parameter. So in this case, it would be perfectly fine to have `closeCh` captured within the state, and then listen to it to cancel the locally created context instance. Though I must agree that it is a tempting to put `context` to the state :slightly_smiling_face: 

Here are the reasons why it is considered a bad practice.
> There are a few reasons why it's generally considered bad practice to add a `context.Context `reference to a state object in Golang:
> 
> It violates the principle of separation of concerns: Adding a `context.Context` reference to a state object means that your state object is now responsible for both managing its own internal state and handling request-scoped values. This can make your code harder to understand and reason about.
> 
> It can lead to coupling: Storing a `context.Context` reference in your state object means that your state object is now tightly coupled to the request context. This can make it harder to reuse your state object in other contexts, since it now depends on a specific `context.Context` implementation.
> 
> It can make testing more difficult: Including a `context.Context `reference in your state object can make it harder to test your code, since you now need to provide a valid `context.Context` for your tests to run. This can make your tests more complex and harder to maintain.
> 
> It can introduce subtle bugs: Storing a `context.Context` reference in your state object can lead to subtle bugs if you're not careful. For example, if you pass a reference to a `context.Context` object to a goroutine, that goroutine will continue to run even if the original request context is canceled. This can lead to resource leaks and other issues.
> 
> Overall, while there may be some cases where it makes sense to include a `context.Context` reference in a state object, it's generally considered bad practice to do so. Instead, you should look for alternative ways to manage your state and handle request-scoped values, such as passing values explicitly through function arguments or using other data structures to manage your state.

Reply:
reverted https://github.com/0xPolygon/polygon-edge/pull/1364/commits/e4fa5d62fac2498a5ee0740964dc1aa6f5a277c1